dashboard/test/controllers/registrations_controller/set_email_test.rb

```diff
@@ -0,0 +1,302 @@
+#
+class SetEmailTest < ActionController::TestCase
+  setup do
+    @controller = RegistrationsController.new
+
+    # stub properties so we don't try to hit pegasus db
+    Properties.stubs(:get).returns nil
+
+    # This is an AJAX-first route
+    request.headers['HTTP_ACCEPT'] = "application/json"
+  end
+
+  test "set email without user param returns 400 BAD REQUEST" do
+    student = create :student
+    sign_in student
+    assert_does_not_create(User) do
+      post :set_email, params: {}
+    end
+    assert_response :bad_request
+  end
+
+  test "set email with utf8mb4 in email fails" do
+    student = create :student
+    sign_in student
+
+    # don't ask the db for existing panda emails
+    User.expects(:find_by_email_or_hashed_email).never
+
+    assert_does_not_create(User) do
+      post :set_email, params: {
+        user: {email: "#{panda_panda}@panda.xx", current_password: '00secret'}
+      }
+    end
+
+    assert_response :unprocessable_entity
+    assert_equal ['Email is invalid'], assigns(:user).errors.full_messages
+    assert_equal response.body, {email: ['Email is invalid']}.to_json
+  end
+
+  test "update student with client side hashed email" do
+    student = create :student, birthday: '1981/03/24', password: 'whatev'
+    sign_in student
+
+    post :set_email, params: {
+      user: {
+        email: '',
+        hashed_email: User.hash_email('hidden@email.com'),
+        current_password: 'whatev' # need this to change email
+      }
+    }
+
+    assert_response :success
+    assert_equal '', assigns(:user).email
+    assert_equal User.hash_email('hidden@email.com'), assigns(:user).hashed_email
+  end
+
+  test "update over 13 student with plaintext email" do
+    student = create :student, birthday: '1981/03/24', password: 'whatev'
+    sign_in student
+
+    post :set_email, params: {
+      user: {
+        email: 'hashed@email.com',
+        current_password: 'whatev' # need this to change email
+      }
+    }
+
+    assert_response :success
+    assert_equal '', assigns(:user).email
+    assert_equal User.hash_email('hashed@email.com'), assigns(:user).hashed_email
+  end
+
+  test 'update rejects unwanted parameters' do
+    user = create :user, name: 'non-admin'
+    sign_in user
+    post :set_email, params: {user: {admin: true}}
+    assert_response :success
+
+    user.reload
+    refute user.admin
+  end
+
+  # The next several tests explore profile changes for users with or without
+  # passwords.  Examples of users without passwords are users that authenticate
+  # via oauth (a third-party account), or students with a picture password.
+
+  # Tech debt note:
+  # These tests make multiple controller calls per-test, which is not fully supported as per http://api.rubyonrails.org/v4.2/classes/ActionController/TestCase.html
+  # Currently this is worked around by calling current_user.reload in registrations_controller.rb, but ideally
+  # these tests should be fixed up to avoid this issue.
```

Comment:
Do we have a backlog to track this kind of work somewhere or is in-code documentation enough?

Reply:
I [copied this note from registrations_controller_test.rb](https://github.com/code-dot-org/code-dot-org/blob/ee7f77feafc9453d379bfe9a83f853c9da51f0cc/dashboard/test/controllers/registrations_controller_test.rb#L489-L492) because these tests inherit the same issue.  I think the cost of this tech debt is low enough that I don't feel a need to track it elsewhere - but if we care about it, I could break apart the offending unit tests now and remove the workaround from the new route.  Thoughts?

Reply:
Agree - doesn't feel high priority.  I was just curious.

Reply:
Updated!  Thanks for pointing this out!